contracts/cw-ics721-bridge/src/ibc.rs

```diff
@@ -0,0 +1,466 @@
+                            token_ids,
+                        }))?,
+                        funds: vec![],
+                    },
+                    ACK_AND_DO_NOTHING,
+                ));
+            }
+            if let Some(InstantiateAndMintInfo {
+                class_id,
+                token_ids,
+                token_uris,
+            }) = self.iandm
+            {
+                messages.push(SubMsg::reply_always(
+                    WasmMsg::Execute {
+                        contract_addr: contract.into_string(),
+                        msg: to_binary(&ExecuteMsg::Callback(CallbackMsg::DoInstantiateAndMint {
+                            class_id,
+                            class_uri,
+                            token_ids,
+                            token_uris,
+                            receiver: receiver.into_string(),
+                        }))?,
+                        funds: vec![],
+                    },
+                    ACK_AND_DO_NOTHING,
+                ))
+            }
+            Ok(messages)
+        }
+    }
+
+    let data: NonFungibleTokenPacketData = from_binary(&packet.data)?;
+    data.validate()?;
+
+    let local_class_id = try_pop_source_prefix(&packet.src, &data.class_id);
+    let receiver = deps.api.addr_validate(&data.receiver)?;
+    let token_count = data.token_ids.len();
+
+    // Say we're connected to three identical chains A, B, and C. For
+    // each of these chains call the local channel ID `C` and the
+    // local port `P`. After taking the path (A -> B -> C) the class
+    // ID on C is `P/C/P/C/P/C/contract_address`.
+    //
+    // Now, lets say the next hop we take is from (C -> A). A receives
+    // a packet with prefix `P/C`. According to the logic on the spec,
+    // it would recognize this as its prefix and attempt to release
+    // its local version of the NFT (from the prefix alone, it seems
+    // like this has previously been transfered away!). This attempt
+    // to release fails though as there has never been a transfer from
+    // (A -> C)!
+    //
+    // What do we need to do instead? Before attempting the transfer,
+    // we need to verify that the incoming NFT has previously been
+    // transfered out. If it has not, we should not attempt the
+    // transfer and instead (correctly) treat it as a new NFT that we
+    // have not seen before and create a new local cw721 contract.
+    let messages = data
+        .token_ids
+        .into_iter()
+        .zip(data.token_uris.into_iter())
+        .try_fold(
+            Vec::<Action>::with_capacity(token_count),
+            |mut messages, (token, token_uri)| -> StdResult<_> {
+                if let Some(local_class_id) = local_class_id {
+                    let key = (local_class_id.to_string(), token.clone());
+                    let outgoing_channel =
+                        OUTGOING_CLASS_TOKEN_TO_CHANNEL.may_load(deps.storage, key.clone())?;
```

Comment:
Given transfer A -> B

This is being called on chain B, so here `OUTGOING_CLASS_TOKEN_TO_CHANNEL` should only contain NFTs that are sent away from chain B to some other chain, this code doesn't have access to `OUTGOING_CLASS_TOKEN_TO_CHANNEL` store that is on chain A (The one that holds the sent away NFT data.

What am i missing here?

Reply:
when we're receiving a NFT from another chain we need to check if we have previously sent that NFT out to the sending chain. if we have, the NFT is returning to us so we should transfer the original NFT to the receiver and we expect that the sending chain will burn the debt voucher (as it has been repaid) on their side upon receiving an ack.

here's where that gets tricky:

we can not actually trust the class ID field of the incoming NFT alone to determine if the NFT in question is returning. IBC is permissionless! anyone can make any "chain" and send us any data. as such we need to check:

1. does the class ID of the incoming NFT begin with the class ID for the channel it is coming in on.
2. has the NFT coming in previously been sent out on the channel this NFT is coming in on.

if _both_ (1) and (2) are true then we can release the NFT to the receiver.

what if only (1) is true?

in that case, one of two things are possible:

1. the incoming NFT is coming from a chain with the same channel and port ID as us (the chain is a duplicate of us).
2. the incoming NFT is invalid and it's spoofed data.

the thing is: we can't tell the difference between these two! as such, if (2) isn't true above we fall through to our logic that handles us receiving a new NFT and mint one for the receiver.

we could prevent this fall-through case by just erroring, though, i personally feel weird about explicitly disallowing a NFT path that is allowed by the spec, especially when doing so would require additional logic in the contract.

does this make sense? would love more thoughts / questions here if this is unclear.